compiler/rustc_codegen_llvm/src/llvm_util.rs

```diff
@@ -118,14 +118,20 @@ unsafe fn configure_llvm(sess: &Session) {
 
     llvm::LLVMInitializePasses();
 
-    for plugin in &sess.opts.debugging_opts.llvm_plugins {
-        let path = Path::new(plugin);
-        let res = DynamicLibrary::open(path);
-        match res {
-            Ok(_) => debug!("LLVM plugin loaded succesfully {} ({})", path.display(), plugin),
-            Err(e) => bug!("couldn't load plugin: {}", e),
+    let use_new_llvm_pm_plugin_register =
+        sess.opts.debugging_opts.new_llvm_pass_manager.unwrap_or(false);
```

Comment:
If `new_llvm_pass_manager` returns `None`, the new pass manager is used when the llvm version is at least 13 and the target is not s390x. See https://github.com/rust-lang/rust/blob/8b09ba6a5d5c644fe0f1c27c7f9c80b334241707/compiler/rustc_codegen_llvm/src/back/write.rs#L397-L410

Reply:
Yes that's what i was mentioning at the end of https://github.com/rust-lang/rust/pull/91125#issuecomment-976710806.

What i'm proposing is to condition the new PM plugin loading only on the explicit `new_llvm_pass_manager=yes` and not the actual usage of the new PM later. Since there are exceptions and a default value hardcoded, we can't trust the option to reflect reality.

If we do want to condition on the actual usage of the new PM then i see two potential issues:
- current users of `llvm-plugins` when updating to a llvm13 based rust will trigger PassPlugin::Load on their legacy plugins since the new PM is enabled by default. They will need to explicitly set `new_llvm_pass_manager=false`
- i'm not sure how to use `should_use_new_llvm_pass_manager` in llvm_util.rs, maybe just move it to llvm_util.rs to make it more widely available ?

Reply:
@bjorn3 do you think the current code is acceptable regarding the explicit usage of `new_llvm_pass_manager` or i should
propose the alternative described in my last reply ?

Reply:
Sorry for the late reply. Missed your previous comment.

> current users of llvm-plugins when updating to a llvm13 based rust will trigger PassPlugin::Load on their legacy plugins since the new PM is enabled by default. They will need to explicitly set new_llvm_pass_manager=false

At least in that case you will get an error message rather than silently ignoring it, right?

> i'm not sure how to use should_use_new_llvm_pass_manager in llvm_util.rs, maybe just move it to llvm_util.rs to make it more widely available ?

Makes sense I think.

> do you think the current code is acceptable regarding the explicit usage of new_llvm_pass_manager or i should
propose the alternative described in my last reply ?

I think using `should_use_new_llvm_pass_manager` is better. I don't strongly object to the current code though.

Reply:
> Sorry for the late reply. Missed your previous comment.
> 
No problem, i should have pinged you earlier but i was busy on other tasks anyway.
> > current users of llvm-plugins when updating to a llvm13 based rust will trigger PassPlugin::Load on their legacy plugins since the new PM is enabled by default. They will need to explicitly set new_llvm_pass_manager=false
> 
> At least in that case you will get an error message rather than silently ignoring it, right?
> 
Indeed, in that case  `cargo build` fails with an error and i can see original LLVM error message directly in cargo non verbose logs like this: `Plugin entry point not found in '/tmp/libLLVMPlugin.so'. Is this a legacy plugin?`
> > i'm not sure how to use should_use_new_llvm_pass_manager in llvm_util.rs, maybe just move it to llvm_util.rs to make it more widely available ?
> 
> Makes sense I think.
Ok let's try it.
> 
> > do you think the current code is acceptable regarding the explicit usage of new_llvm_pass_manager or i should
> > propose the alternative described in my last reply ?
> 
> I think using `should_use_new_llvm_pass_manager` is better. I don't strongly object to the current code though.
Ok thank you for your feedback, i'll try doing it this way and see how that goes.